### PR TITLE
[IM] Product Variation Support

### PR DESCRIPTION
**Pull request by Cdok**
# [IM] Product Variation Support

**The SDK PR https://github.com/mobify/progressive-web-sdk/pull/715 needs to be merged & released before merging this scaffold branch**

 **JIRA**: https://mobify.atlassian.net/browse/WEBDATA-88

## Todos

- [x] `Update Cart` functionality on merlin's variants
- [x] `Add to Cart` functionality on merlin's variants
- [x] ~~Refactor to avoid changing `addToCart` command parameters~~
- [x] Fix runtypes for new optional variationCategory property
- [x] Ensure preselect behavior exists as it did previously on SFCC
- [x] Test for regressions across both connectors
- [x] Test for regressions on non-variant products

## Changes
- Captures the variation data from products on Merlin's potions
- Format data to match the variation schema on SFCC connector
- Update addToCart / submitCartForm to allow adding to cart
- Update `cart-product-list` to display selected variants in cart

## How to test-drive this PR

```
cd progressive-web-sdk
git checkout im-add-to-cart-variants
npm link
npm install
npm run dev:build
cd ../platform-scaffold/web
git checkout product-variants-on-merlins
npm link progressive-web-sdk
npm run dev
```

- Use Merlin's Connector
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- Add Eye of Newt with variations to cart
- Attempt updating the cart
- Add a different variation to cart
- Ensure both variants are showing up as separate cart items

### CODE REVIEW

File: web/app/integration-manager/cart/commands.js [hunk 1]
```diff
@@ -56,8 +56,8 @@ const dispatchCartAnalytics = (action, dispatch, getState, id) => {
  * @param productId {string} The product's ID
  * @param quantity {number} The quantity to add
  */
```
Comment (jnurse) [inline web/app/integration-manager/cart/commands.js]: Can you update the JSDoc comments to reflect the new function parameters? And maybe include a note in the description that not all connectors will require the variant object?

File: web/app/integration-manager/cart/commands.js [hunk 2]
```diff
@@ -82,8 +82,8 @@ export const removeFromCart = (ItemId) => (dispatch, getState) => {
  * @param quantity {number} The new quantity
  * @param productId {string} The new product ID to replace cart item
  */
```
Comment (jnurse) [inline web/app/integration-manager/cart/commands.js]: Same as above, can you update the JSDoc comments?

**Comment by Cdok** [comment]
Todos:
- [ ] Preselect first variant on **Merlin's**
- [ ] Preselect item from EDIT in cart on **Merlin's**
- [ ] Display variant data in cart on **SFCC** (`getCart` doesn't return what we need)
- [ ] Update `onVariationChange` to update product data (price, picture, etc) on **Merlin's**

**Comment by Cdok** [comment]
@jeremywiebe The cart / mini cart should be resolved now, there was a bit of a refactor happening on Friday that I've just finished cleaning up.

I think documentation is a good idea as well, as this actually touches quite a bit of how variations were implemented on SFCC as well.

File: web/app/connectors/_merlins-connector/cart/commands.js [hunk 3]
```diff
@@ -52,7 +52,17 @@ export const getCart = () => (dispatch) => {
...
-export const addToCart = (productId, quantity) => (dispatch, getState) => {
+/**
+ * @function addToCart
+ * @param {String} productId The product's ID
+ * @param {Number} quantity The quantity to add
+ * @param {Object} variant The variant object (required for this connector)
+ * @param {Number} variant.id the unique product combination's ID
+ * @param {Object} variant.values map of attribute slugs & selected values, (ex. color: 8, size: 6)
+ * @param {Object} variant.attributeIds map of options to option IDs, (ex. color: 90)
+ * where optionID is Magento's numerical representation of the attribute
```
Comment (jeremywiebe) [inline web/app/connectors/_merlins-connector/cart/commands.js]: I think it'd make the `variant.values` and `variant.attributeIds` a bit clearer if you had a small [example](http://usejsdoc.org/tags-example.html) following this param list.

File: web/app/connectors/_sfcc-connector/cart/commands.js [hunk 4]
```diff
@@ -18,6 +18,11 @@ const addToCartRequest = (productId, quantity, basketId) => {
     return makeApiJsonRequest(`/baskets/${basketId}/items`, requestBody, {method: 'POST'})
 }
 
+/**
+ * @function addToCart
+ * @param {String} productId The product's ID
```
Comment (jeremywiebe) [inline web/app/connectors/_sfcc-connector/cart/commands.js]: Let's add a note here that if the product supports variants, this is the product variant ID and cannot be the master product ID.

File: web/app/containers/product-details/actions.js [hunk 5]
```diff
@@ -48,29 +48,49 @@ export const goToCheckout = () => (dispatch, getState) => {
...
 })
 
+/**
+ * @param {Object} formValues form values of selected variant options
+ * {
+ *	"color": "8",
+ *	"size": "6"
+ * }
+ * @param {Array} variants all possible variants, array of {Objects}, each object has a values key
+ * the values in this key will have the same shape / data as the above formValues if they match
```
Comment (jeremywiebe) [inline web/app/containers/product-details/actions.js]: Can we add a period at the end of the previous line and start this with a capital? I had to read it a few times to realize the line break was also a sentence break.

**Pull request merged**